examples/pytorch/graphsage/experimental/train_dist_transductive.py

```diff
@@ -0,0 +1,319 @@
+                shuffle=False,
+                drop_last=False)
+
+            for blocks in tqdm.tqdm(dataloader):
+                block = blocks[0].to(device)
+                input_nodes = block.srcdata[dgl.NID]
+                output_nodes = block.dstdata[dgl.NID]
+                h = x[input_nodes].to(device)
+                h_dst = h[:block.number_of_dst_nodes()]
+                h = layer(block, (h, h_dst))
+                if l != len(self.layers) - 1:
+                    h = self.activation(h)
+                    h = self.dropout(h)
+
+                y[output_nodes] = h.cpu()
+
+            x = y
+            g.barrier()
+        return y
+
+def initializer(shape, dtype):
+    arr = th.zeros(shape, dtype=dtype)
+    arr.uniform_(-1, 1)
+    return arr
+
+class DistEmb(nn.Module):
+    def __init__(self, num_nodes, emb_size, dgl_sparse_emb=False, dev_id='cpu'):
+        super().__init__()
+        self.dev_id = dev_id
+        self.emb_size = emb_size
+        self.dgl_sparse_emb = dgl_sparse_emb
+        if dgl_sparse_emb:
+            self.sparse_emb = NodeEmbedding(num_nodes, emb_size, name='sage', init_func=initializer)
+        else:
+            self.sparse_emb = th.nn.Embedding(num_nodes, emb_size, sparse=True)
+            nn.init.uniform_(self.sparse_emb.weight, -1.0, 1.0)
+
+    def forward(self, idx):
+        # embeddings are stored in cpu
+        idx = idx.cpu()
+        if self.dgl_sparse_emb:
+            return self.sparse_emb(idx, device=self.dev_id)
+        else:
+            return self.sparse_emb(idx).to(self.dev_id)
+
+def load_embs(standalone, emb_layer, g):
+    nodes = dgl.distributed.node_split(np.arange(g.number_of_nodes()),
+                                       g.get_partition_book(), force_even=True)
+    x = dgl.distributed.DistTensor(
+        (g.number_of_nodes(),
+         emb_layer.module.emb_size \
+            if isinstance(emb_layer, th.nn.parallel.DistributedDataParallel) \
+            else emb_layer.emb_size),
+        th.float32, 'eval_embs',
+        persistent=True)
+    num_nodes = nodes.shape[0]
+    for i in range((num_nodes + 1023) // 1024):
+        idx = nodes[i * 1024: (i+1) * 1024 \
+                    if (i+1) * 1024 < num_nodes \
+                    else num_nodes]
+        embeds = emb_layer(idx).cpu()
+        x[idx] = embeds
+
+    if not standalone:
+        g.barrier()
+
+    return x
+
+def evaluate(standalone, model, emb_layer, g, labels, val_nid, test_nid, batch_size, device):
+    """
+    Evaluate the model on the validation set specified by ``val_nid``.
+    g : The entire graph.
+    inputs : The features of all the nodes.
+    labels : The labels of all the nodes.
+    val_nid : the node Ids for validation.
+    batch_size : Number of nodes to compute at the same time.
+    device : The GPU device to evaluate on.
+    """
+    model.eval()
+    emb_layer.eval()
+    with th.no_grad():
+        inputs = load_embs(standalone, emb_layer, g)
```

Comment:
can we just call `emb_layer.emb_tensor` for `NodeEmbedding` or something like that for Pytorch embedding?

Reply:
I think it is better not to access emb_layer.emb_tensor directly.